tests/kdb/testkdb_allplugins.cpp

```diff
@@ -43,6 +44,11 @@ std::vector<std::string> getAllPlugins ()
 	// Valgrind reports memory leaks for the `semlock` plugin on Debian Unstable: http://issues.libelektra.org/2113
 	plugins.erase (std::remove (plugins.begin (), plugins.end (), "semlock"), plugins.end ());
 
+#ifdef ENABLE_ASAN
```

Comment:
I wonder if we are able to auto-generate this list here by checking which plugins have memleak. (of course disable only for asan adds extra complexity)

Reply:
That would certainly make things much easier for plugin developers. However, as far as I can tell, we currently do not provide an easy way to check the the status of a plugin inside CMake (or C++) code.

Reply:
In CMake it is a bit hacky (but we also parse the README.md in CMake), in C++ we have the lookupInfo in Plugin of libtools. Libtools is used in this file anyway, so it should not be too much trouble to add this feature. It is also not trivial, however.

Reply:
I added the following code:

```cpp
plugins.erase (std::remove (plugins.begin (), plugins.end (), "jni"), plugins.end ());

for (std::string plugin : plugins)
{
	try
	{
		auto spec = mpd.lookupInfo (PluginSpec (plugin), "status");
	}
	catch (std::exception const & e)
	{
		std::cerr << "Unable to determine status of plugin “" << plugin << "”" << std::endl;
	}
}
```

at the start of `getAllPlugins`. Since this update already produces memory leaks:

```
 ==13485==ERROR: LeakSanitizer: detected memory leaks
192: 
192: Direct leak of 1480 byte(s) in 1 object(s) allocated from:
192:     #0 0x1073cec63 in wrap_malloc (libclang_rt.asan_osx_dynamic.dylib:x86_64h+0x56c63)
192:     #1 0x7fff769dc2bb in _duplocale (libsystem_c.dylib:x86_64+0xb2bb)
192:     #2 0x7fff769dc574 in newlocale (libsystem_c.dylib:x86_64+0xb574)
192:     #3 0x10ab40589  (<unknown module>)
192:     #4 0x10ab40751  (<unknown module>)
192:     #5 0x10ab23114  (<unknown module>)
192:     #6 0x107292790 in elektraPluginOpen plugin.c:292
192:     #7 0x106e79ec7 in kdb::tools::Plugin::Plugin(kdb::tools::PluginSpec const&, kdb::KeySet&) plugin.cpp:39
192:     #8 0x106e79752 in kdb::tools::Modules::load(kdb::tools::PluginSpec const&) modules.cpp:49
192:     #9 0x106e794ec in kdb::tools::Modules::load(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, kdb::KeySet const&) modules.cpp:44
192:     #10 0x106edb4ea in kdb::tools::ModulesPluginDatabase::lookupInfo(kdb::tools::PluginSpec const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) const plugindatabase.cpp:219
192:     #11 0x106988592 in getAllPlugins() testkdb_allplugins.cpp:38
192:     #12 0x10699ef68 in gtest_AllPluginsAllPlugins_EvalGenerator_() testkdb_allplugins.cpp:110
192:     #13 0x1069b10e4 in testing::internal::ParameterizedTestCaseInfo<AllPlugins>::RegisterTests() gtest-param-util.h:548
192:     #14 0x106a30203 in testing::internal::ParameterizedTestCaseRegistry::RegisterTests() gtest-param-util.h:708
192:     #15 0x106a2fef4 in testing::internal::UnitTestImpl::RegisterParameterizedTests() gtest.cc:2632
192:     #16 0x106a76af8 in testing::internal::UnitTestImpl::PostFlagParsingInit() gtest.cc:4857
192:     #17 0x106a831dc in void testing::internal::InitGoogleTestImpl<char>(int*, char**) gtest.cc:5772
192:     #18 0x1069ddb6f in main gtest_main.cc:36
192:     #19 0x7fff76982014 in start (libdyld.dylib:x86_64+0x1014)
192: 
192: Direct leak of 1480 byte(s) in 1 object(s) allocated from:
192:     #0 0x1073cec63 in wrap_malloc (libclang_rt.asan_osx_dynamic.dylib:x86_64h+0x56c63)
192:     #1 0x7fff769dc2bb in _duplocale (libsystem_c.dylib:x86_64+0xb2bb)
192:     #2 0x7fff769dc574 in newlocale (libsystem_c.dylib:x86_64+0xb574)
192:     #3 0x10ab60589  (<unknown module>)
192:     #4 0x10ab60751  (<unknown module>)
192:     #5 0x10ab43114  (<unknown module>)
192:     #6 0x107292790 in elektraPluginOpen plugin.c:292
192:     #7 0x106e79ec7 in kdb::tools::Plugin::Plugin(kdb::tools::PluginSpec const&, kdb::KeySet&) plugin.cpp:39
192:     #8 0x106e79752 in kdb::tools::Modules::load(kdb::tools::PluginSpec const&) modules.cpp:49
192:     #9 0x106e794ec in kdb::tools::Modules::load(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, kdb::KeySet const&) modules.cpp:44
192:     #10 0x106edb4ea in kdb::tools::ModulesPluginDatabase::lookupInfo(kdb::tools::PluginSpec const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) const plugindatabase.cpp:219
192:     #11 0x106988592 in getAllPlugins() testkdb_allplugins.cpp:38
192:     #12 0x10699ef68 in gtest_AllPluginsAllPlugins_EvalGenerator_() testkdb_allplugins.cpp:110
192:     #13 0x1069b10e4 in testing::internal::ParameterizedTestCaseInfo<AllPlugins>::RegisterTests() gtest-param-util.h:548
192:     #14 0x106a30203 in testing::internal::ParameterizedTestCaseRegistry::RegisterTests() gtest-param-util.h:708
192:     #15 0x106a2fef4 in testing::internal::UnitTestImpl::RegisterParameterizedTests() gtest.cc:2632
192:     #16 0x106a76af8 in testing::internal::UnitTestImpl::PostFlagParsingInit() gtest.cc:4857
192:     #17 0x106a831dc in void testing::internal::InitGoogleTestImpl<char>(int*, char**) gtest.cc:5772
192:     #18 0x1069ddb6f in main gtest_main.cc:36
192:     #19 0x7fff76982014 in start (libdyld.dylib:x86_64+0x1014)
192: 
192: Direct leak of 144 byte(s) in 6 object(s) allocated from:
192:     #0 0x1073cec63 in wrap_malloc (libclang_rt.asan_osx_dynamic.dylib:x86_64h+0x56c63)
192:     #1 0x10b4f11b9  (<unknown module>)
192:     #2 0x10b4f14d5  (<unknown module>)
192:     #3 0x10b551ace  (<unknown module>)
192:     #4 0x10b4f015c  (<unknown module>)
192:     #5 0x10b4f0921  (<unknown module>)
192:     #6 0x10b4ed9eb  (<unknown module>)
192:     #7 0x10b4aea15  (<unknown module>)
192:     #8 0x10b4bf725  (<unknown module>)
192:     #9 0x107292790 in elektraPluginOpen plugin.c:292
192:     #10 0x106e79ec7 in kdb::tools::Plugin::Plugin(kdb::tools::PluginSpec const&, kdb::KeySet&) plugin.cpp:39
192:     #11 0x106e79752 in kdb::tools::Modules::load(kdb::tools::PluginSpec const&) modules.cpp:49
192:     #12 0x106e794ec in kdb::tools::Modules::load(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, kdb::KeySet const&) modules.cpp:44
192:     #13 0x106edb4ea in kdb::tools::ModulesPluginDatabase::lookupInfo(kdb::tools::PluginSpec const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) const plugindatabase.cpp:219
192:     #14 0x106988592 in getAllPlugins() testkdb_allplugins.cpp:38
192:     #15 0x10699ef68 in gtest_AllPluginsAllPlugins_EvalGenerator_() testkdb_allplugins.cpp:110
192:     #16 0x1069b10e4 in testing::internal::ParameterizedTestCaseInfo<AllPlugins>::RegisterTests() gtest-param-util.h:548
192:     #17 0x106a30203 in testing::internal::ParameterizedTestCaseRegistry::RegisterTests() gtest-param-util.h:708
192:     #18 0x106a2fef4 in testing::internal::UnitTestImpl::RegisterParameterizedTests() gtest.cc:2632
192:     #19 0x106a76af8 in testing::internal::UnitTestImpl::PostFlagParsingInit() gtest.cc:4857
192:     #20 0x106a831dc in void testing::internal::InitGoogleTestImpl<char>(int*, char**) gtest.cc:5772
192:     #21 0x1069ddb6f in main gtest_main.cc:36
192:     #22 0x7fff76982014 in start (libdyld.dylib:x86_64+0x1014)
192: 
192: Direct leak of 144 byte(s) in 6 object(s) allocated from:
192:     #0 0x1073cec63 in wrap_malloc (libclang_rt.asan_osx_dynamic.dylib:x86_64h+0x56c63)
192:     #1 0x10b2c71b9  (<unknown module>)
192:     #2 0x10b2c74d5  (<unknown module>)
192:     #3 0x10b327ace  (<unknown module>)
192:     #4 0x10b2c615c  (<unknown module>)
192:     #5 0x10b2c6921  (<unknown module>)
192:     #6 0x10b2c39eb  (<unknown module>)
192:     #7 0x10b284a15  (<unknown module>)
192:     #8 0x10b295725  (<unknown module>)
192:     #9 0x107292790 in elektraPluginOpen plugin.c:292
192:     #10 0x106e79ec7 in kdb::tools::Plugin::Plugin(kdb::tools::PluginSpec const&, kdb::KeySet&) plugin.cpp:39
192:     #11 0x106e79752 in kdb::tools::Modules::load(kdb::tools::PluginSpec const&) modules.cpp:49
192:     #12 0x106e794ec in kdb::tools::Modules::load(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, kdb::KeySet const&) modules.cpp:44
192:     #13 0x106edb4ea in kdb::tools::ModulesPluginDatabase::lookupInfo(kdb::tools::PluginSpec const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) const plugindatabase.cpp:219
192:     #14 0x106988592 in getAllPlugins() testkdb_allplugins.cpp:38
192:     #15 0x10699ef68 in gtest_AllPluginsAllPlugins_EvalGenerator_() testkdb_allplugins.cpp:110
192:     #16 0x1069b10e4 in testing::internal::ParameterizedTestCaseInfo<AllPlugins>::RegisterTests() gtest-param-util.h:548
192:     #17 0x106a30203 in testing::internal::ParameterizedTestCaseRegistry::RegisterTests() gtest-param-util.h:708
192:     #18 0x106a2fef4 in testing::internal::UnitTestImpl::RegisterParameterizedTests() gtest.cc:2632
192:     #19 0x106a76af8 in testing::internal::UnitTestImpl::PostFlagParsingInit() gtest.cc:4857
192:     #20 0x106a831dc in void testing::internal::InitGoogleTestImpl<char>(int*, char**) gtest.cc:5772
192:     #21 0x1069ddb6f in main gtest_main.cc:36
192:     #22 0x7fff76982014 in start (libdyld.dylib:x86_64+0x1014)
192: 
192: Indirect leak of 48 byte(s) in 6 object(s) allocated from:
192:     #0 0x1073cec63 in wrap_malloc (libclang_rt.asan_osx_dynamic.dylib:x86_64h+0x56c63)
192:     #1 0x10b2c71b9  (<unknown module>)
Errors while running CTest
make: *** [Makefile:75: shell] Error 8
192:     #2 0x10b2c74d5  (<unknown module>)
192:     #3 0x10b327adb  (<unknown module>)
192:     #4 0x10b2c615c  (<unknown module>)
192:     #5 0x10b2c6921  (<unknown module>)
192:     #6 0x10b2c39eb  (<unknown module>)
192:     #7 0x10b284a15  (<unknown module>)
192:     #8 0x10b295725  (<unknown module>)
192:     #9 0x107292790 in elektraPluginOpen plugin.c:292
192:     #10 0x106e79ec7 in kdb::tools::Plugin::Plugin(kdb::tools::PluginSpec const&, kdb::KeySet&) plugin.cpp:39
192:     #11 0x106e79752 in kdb::tools::Modules::load(kdb::tools::PluginSpec const&) modules.cpp:49
192:     #12 0x106e794ec in kdb::tools::Modules::load(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, kdb::KeySet const&) modules.cpp:44
192:     #13 0x106edb4ea in kdb::tools::ModulesPluginDatabase::lookupInfo(kdb::tools::PluginSpec const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) const plugindatabase.cpp:219
192:     #14 0x106988592 in getAllPlugins() testkdb_allplugins.cpp:38
192:     #15 0x10699ef68 in gtest_AllPluginsAllPlugins_EvalGenerator_() testkdb_allplugins.cpp:110
192:     #16 0x1069b10e4 in testing::internal::ParameterizedTestCaseInfo<AllPlugins>::RegisterTests() gtest-param-util.h:548
192:     #17 0x106a30203 in testing::internal::ParameterizedTestCaseRegistry::RegisterTests() gtest-param-util.h:708
192:     #18 0x106a2fef4 in testing::internal::UnitTestImpl::RegisterParameterizedTests() gtest.cc:2632
192:     #19 0x106a76af8 in testing::internal::UnitTestImpl::PostFlagParsingInit() gtest.cc:4857
192:     #20 0x106a831dc in void testing::internal::InitGoogleTestImpl<char>(int*, char**) gtest.cc:5772
192:     #21 0x1069ddb6f in main gtest_main.cc:36
192:     #22 0x7fff76982014 in start (libdyld.dylib:x86_64+0x1014)
192: 
192: Indirect leak of 48 byte(s) in 6 object(s) allocated from:
192:     #0 0x1073cec63 in wrap_malloc (libclang_rt.asan_osx_dynamic.dylib:x86_64h+0x56c63)
192:     #1 0x10b4f11b9  (<unknown module>)
192:     #2 0x10b4f14d5  (<unknown module>)
192:     #3 0x10b551adb  (<unknown module>)
192:     #4 0x10b4f015c  (<unknown module>)
192:     #5 0x10b4f0921  (<unknown module>)
192:     #6 0x10b4ed9eb  (<unknown module>)
192:     #7 0x10b4aea15  (<unknown module>)
192:     #8 0x10b4bf725  (<unknown module>)
192:     #9 0x107292790 in elektraPluginOpen plugin.c:292
192:     #10 0x106e79ec7 in kdb::tools::Plugin::Plugin(kdb::tools::PluginSpec const&, kdb::KeySet&) plugin.cpp:39
192:     #11 0x106e79752 in kdb::tools::Modules::load(kdb::tools::PluginSpec const&) modules.cpp:49
192:     #12 0x106e794ec in kdb::tools::Modules::load(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, kdb::KeySet const&) modules.cpp:44
192:     #13 0x106edb4ea in kdb::tools::ModulesPluginDatabase::lookupInfo(kdb::tools::PluginSpec const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) const plugindatabase.cpp:219
192:     #14 0x106988592 in getAllPlugins() testkdb_allplugins.cpp:38
192:     #15 0x10699ef68 in gtest_AllPluginsAllPlugins_EvalGenerator_() testkdb_allplugins.cpp:110
192:     #16 0x1069b10e4 in testing::internal::ParameterizedTestCaseInfo<AllPlugins>::RegisterTests() gtest-param-util.h:548
192:     #17 0x106a30203 in testing::internal::ParameterizedTestCaseRegistry::RegisterTests() gtest-param-util.h:708
192:     #18 0x106a2fef4 in testing::internal::UnitTestImpl::RegisterParameterizedTests() gtest.cc:2632
192:     #19 0x106a76af8 in testing::internal::UnitTestImpl::PostFlagParsingInit() gtest.cc:4857
192:     #20 0x106a831dc in void testing::internal::InitGoogleTestImpl<char>(int*, char**) gtest.cc:5772
192:     #21 0x1069ddb6f in main gtest_main.cc:36
192:     #22 0x7fff76982014 in start (libdyld.dylib:x86_64+0x1014)
192: 
192: -----------------------------------------------------
192: Suppressions used:
192:   count      bytes template
192:       1       3984 *_os_trace*
192: -----------------------------------------------------
192: 
192: SUMMARY: AddressSanitizer: 3344 byte(s) leaked in 26 allocation(s).
```

, I will leave the code as it is for now. I think we have to handle this issue from within CMake.

Reply:
Sorry, my fault. Obviously instantiating the plugin already causes the memleaks. So we also would need some code to suppress any memleaks caused by looking up the contract (because this is not part of the test case but only preparation).